pilot/pkg/serviceregistry/kube/controller/ambient/multicluster.go

```diff
@@ -0,0 +1,549 @@
+		return ptr.Of(krt.WrapClient[T](client, opts.With(
+			krt.WithName("cluster/"+string(c.ID)+"informer/"+name),
+			krt.WithMetadata(krt.Metadata{
+				ClusterKRTMetadataKey: c.ID,
+			}),
+		)...))
+	}
+}
+
+func collectionFromCluster[T controllers.ComparableObject](
+	name string,
+	informerIndex krt.Index[cluster.ID, krt.Collection[T]],
+	opts krt.OptionsBuilder,
+) krt.TransformationSingle[Cluster, krt.Collection[config.ObjectWithCluster[T]]] {
+	return func(ctx krt.HandlerContext, c Cluster) *krt.Collection[config.ObjectWithCluster[T]] {
+		clients := informerIndex.Lookup(c.ID)
+		if len(clients) == 0 {
+			return nil
+		}
+		objWithCluster := krt.NewCollection(clients[0], func(ctx krt.HandlerContext, obj T) *config.ObjectWithCluster[T] {
+			return &config.ObjectWithCluster[T]{ClusterID: c.ID, Object: &obj}
+		}, opts.With(
+			krt.WithName(name+"WithCluster"),
+			krt.WithMetadata(krt.Metadata{
+				ClusterKRTMetadataKey: c.ID,
+			}),
+		)...)
+		return &objWithCluster
+	}
+}
+
+func informerIndexByCluster[T controllers.ComparableObject](
+	informerCollection krt.Collection[krt.Collection[T]],
+) krt.Index[cluster.ID, krt.Collection[T]] {
+	return krt.NewIndex[cluster.ID, krt.Collection[T]](informerCollection, func(col krt.Collection[T]) []cluster.ID {
+		val, ok := col.Metadata()[ClusterKRTMetadataKey]
+		if !ok {
+			log.Warnf("Cluster metadata not set on collection %v", col)
+			return nil
+		}
+		id, ok := val.(cluster.ID)
+		if !ok {
+			log.Warnf("Invalid cluster metadata set on collection %v: %v", col, val)
+			return nil
+		}
+		return []cluster.ID{id}
+	})
+}
+
+func nestedCollectionIndexByCluster[T any](
+	collection krt.Collection[krt.Collection[config.ObjectWithCluster[T]]],
+) krt.Index[cluster.ID, krt.Collection[config.ObjectWithCluster[T]]] {
+	return krt.NewIndex[cluster.ID, krt.Collection[config.ObjectWithCluster[T]]](collection, func(col krt.Collection[config.ObjectWithCluster[T]]) []cluster.ID {
+		val, ok := col.Metadata()[ClusterKRTMetadataKey]
+		if !ok {
+			log.Warnf("Cluster metadata not set on collection %v", col)
+			return nil
+		}
+		id, ok := val.(cluster.ID)
+		if !ok {
+			log.Warnf("Invalid cluster metadata set on collection %v: %v", col, val)
+			return nil
+		}
+		return []cluster.ID{id}
+	})
+}
+
+func mergeServiceInfosWithCluster(localClusterID cluster.ID) func(serviceInfos []config.ObjectWithCluster[model.ServiceInfo]) *config.ObjectWithCluster[model.ServiceInfo] {
+	return func(serviceInfos []config.ObjectWithCluster[model.ServiceInfo]) *config.ObjectWithCluster[model.ServiceInfo] {
+		if len(serviceInfos) == 0 {
+			return nil
+		}
+		if len(serviceInfos) == 1 {
+			return &serviceInfos[0]
+		}
+
+		var vips []*workloadapi.NetworkAddress
+		var sans []string
+		var base config.ObjectWithCluster[model.ServiceInfo]
+		for _, obj := range serviceInfos {
+			if obj.Object == nil {
+				continue
+			}
+
+			vips = append(vips, obj.Object.Service.GetAddresses()...)
```

Comment:
I am not sure this is quite right. In the existing Istiod, we only return service VIPs for the relevant cluster (e.g. `s.ClusterVIPs.GetAddressesFor(node.Metadata.ClusterID)`).

This is going to cause problems when the same VIP is assigned to different services across clusters. We do not require non-overlapping service CIDRs

Reply:
I wondered about that; I don't see how we can possibly merge with the current `model.ServiceInfo` data model though. `model.Service` has ClusterVIPs broken out by service but we don't have that here

Reply:
Actually, a service doesn't have a cluster, so keying on cluster isn't particularly helpful in the data plane. I wonder if addresses is sufficient here since we store the network with the address, thereby preventing collisions (we DO require that vips in the same network are unique)